### DISCUSSION TITLE
Quest Resource Synchronization Challenges

### DISCUSSION THREAD

**Discussion by trevorgowing** [opensrp/fhircore#676] [Ideas]
Hi Team

I have been investigating https://github.com/opensrp/fhircore/issues/642 which was slow sync reported by Matt.
This lead me to creating https://github.com/opensrp/fhircore/issues/668 after a discussion with Maimoona.

As can be seen below we are synchronizing all Patients, CarePlans, QuestionnaireResponses etc as well as the application configuration which currently is represented as a json object stored in a Binary resource.

```kotlin
class QuestApplication : Application(), ConfigurableApplication {
  ...
  override val resourceSyncParams: Map<ResourceType, Map<String, String>>
    get() {
      return mapOf(
        ResourceType.Binary to mapOf("_id" to CONFIG_RESOURCE_IDS),
        ResourceType.CarePlan to mapOf(),
        ResourceType.Patient to mapOf(),
        ResourceType.Questionnaire to buildQuestionnaireFilterMap(),
        ResourceType.QuestionnaireResponse to mapOf(),
        ResourceType.Binary to mapOf()
      )
    }
  ...
```

This is what has caused the slow synchronization since we are synchronizing +- 500 Patients in pages of 20.

We then later filter the resources which we display meaning that we have potentially synchronized resources that we may not need.

In summary I feel that the problem is a misalignment of resource filtering between display and synchronization.

We could introduce resource filters as part of the config but then we are introducing an ordering issue in that we would need to synchronize the config before we synchronize the other resources.

At the moment we schedule the resource synchronization on application create as seen below:

```kotlin
class QuestApplication : Application(), ConfigurableApplication {
  ...
  override fun onCreate() {
    super.onCreate()
    ...
    schedulePeriodicSync()
  }
...
  override fun schedulePeriodicSync() {
    this.runPeriodicSync<QuestFhirSyncWorker>()
  }
```

I propose that we need need to rethink what we sync when.

I think that is makes sense to synchronize the application configuration on application start.
But I think that perhaps it only makes sense to sync the other resources on log in.

**Comment by pld** [comment]
Agree on syncing app config first, esp when we use config to do post-login config stuff, like login-screen this will be important.

On patients, if we're tagging through with an app ID, that'll solve some of this?

Sounds like the batch size is also an issue, like we'd probably want to start w/larger batches and back-off down, or the opposite, oh or maybe use the bulk APIs for sync, I know that's on the SDK roadmap.

**Comment by trevorgowing** [comment]
Logged https://github.com/opensrp/fhircore/issues/668 based on this discussion.
